### PR TITLE
BridgePool Factory

### PR DESCRIPTION
**Pull request by gusjavaz**
BridgePool contracts for deploying instances

### CODE REVIEW

File: bridge/contracts/utils/ImmutableAuth.sol [hunk 1]
```diff
@@ -289,6 +343,87 @@ abstract contract ImmutableLiquidityProviderStaking is ImmutableFactory {
     }
 }
 
+abstract contract ImmutableLocalERC1155BridgePoolV1 is ImmutableFactory {
```
Comment (z-j-lin) [inline bridge/contracts/utils/ImmutableAuth.sol:346]: change this to ImmutableNativeERC1155BridgePoolV1

Comment (gusjavaz) [reply bridge/contracts/utils/ImmutableAuth.sol:346]: Actually removed, belong in bridge-pool-base

File: bridge/test/bridgePool/native-deployment.test.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      bridgePoolChainId,
+      bridgePoolVersion,
+      bridgePoolValue,
+      bridgePoolDeployCode,
+    ]
+  );
+  return { fixture, admin };
+}
+
+beforeEach(async function () {
```
[the remark this answers is not in the document]
Comment (gusjavaz) [reply bridge/test/bridgePool/native-deployment.test.ts]: Indeed that was it. What had me confused was that it worked correctly on local. Thanks!

File: bridge/contracts/libraries/factory/BridgePoolFactoryBase.sol [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            _chainID,
+            poolVersion_
+        );
+        //calculate the address of the pool's logic contract
+        address implementation = _logicAddresses[
+            _getImplementationAddressKey(tokenType_, poolVersion_, native)
+        ];
+        _implementation = implementation;
+        //check if the logic exists for the specified pool
+        //TODO determin if this step is still necessary
```
Comment (Cr0wn-Gh0ul) [inline bridge/contracts/libraries/factory/BridgePoolFactoryBase.sol:137]: delete this comment

File: bridge/contracts/utils/BridgePoolAddressUtil.sol [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                bytes.concat(
+                    keccak256(abi.encodePacked(tokenContractAddr_)),
+                    keccak256(abi.encodePacked(tokenType_)),
+                    keccak256(abi.encodePacked(chainID_)),
+                    keccak256(abi.encodePacked(version_))
+                )
+            );
+    }
+
+    function getBridgePoolAddress(bytes32 bridgePoolSalt_, address bridgeFactory_)
```
Comment (Cr0wn-Gh0ul) [inline bridge/contracts/utils/BridgePoolAddressUtil.sol:36]: add test

**Review by Cr0wn-Gh0ul — CHANGES_REQUESTED** [review]
It appears there may have been some mix up commits in this PR.

File: bridge/contracts/libraries/factory/BridgePoolFactoryBase.sol [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+pragma solidity ^0.8.16;
+
+import "contracts/utils/ImmutableAuth.sol";
+import "contracts/libraries/errors/BridgePoolFactoryErrors.sol";
+import "contracts/interfaces/IBridgePool.sol";
+import "contracts/utils/BridgePoolAddressUtil.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+abstract contract BridgePoolFactoryBase is ImmutableFactory {
+    enum TokenType {
```
Comment (Cr0wn-Gh0ul) [inline bridge/contracts/libraries/factory/BridgePoolFactoryBase.sol:11]: test upgrading with new enum types (i.e ERC777)

File: bridge/contracts/BridgePoolFactory.sol [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+     * The pools are created as thin proxies (EIP1167) routing to versioned implementations identified by corresponding salt.
+     * @param tokenType_ type of token (1=ERC20, 2=ERC721)
+     * @param ercContract_ address of ERC20 source token contract
+     * @param implementationVersion_ version of BridgePool implementation to use
+     */
+    function deployNewNativePool(
+        uint8 tokenType_,
+        address ercContract_,
+        uint16 implementationVersion_
+    ) public onlyFactoryOrPublicPoolDeploymentEnabled {
```
Comment (Cr0wn-Gh0ul) [inline bridge/contracts/BridgePoolFactory.sol:23]: Does the modifier name really need to be this long?

File: bridge/contracts/utils/BridgePoolAddressUtil.sol [hunk 7]
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: MIT-open-group
+pragma solidity ^0.8.16;
+
+library BridgePoolAddressUtil {
+    enum TokenType {
```
Comment (Cr0wn-Gh0ul) [inline bridge/contracts/utils/BridgePoolAddressUtil.sol:5]: evaluate if really needed here

File: bridge/docs/MP-268:BridgePoolDeployment.puml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+@startuml
```
Comment (Cr0wn-Gh0ul) [inline bridge/docs/MP-268:BridgePoolDeployment.puml:1]: what is this?

**Review by Cr0wn-Gh0ul — CHANGES_REQUESTED** [review]
See comments and suggestions.

File: bridge/docs/MP-268:BridgePoolDeployment.puml (hunk 8 above)
Comment (gusjavaz) [reply bridge/docs/MP-268:BridgePoolDeployment.puml:1]: Attempt to document the development in UML. Removed and moved to Wiki

File: bridge/contracts/libraries/factory/BridgePoolFactoryBase.sol (hunk 5 above)
Comment (gusjavaz) [reply bridge/contracts/libraries/factory/BridgePoolFactoryBase.sol:11]: Added test

**Review by diligentis — APPROVED** [review]
Re-approved for `go-reviewers`

**Review by Cr0wn-Gh0ul — APPROVED** [review]
LGTM

**Pull request merged**